src/infra/core/src/repos/metadata_chain_impl.rs

```diff
@@ -720,14 +631,46 @@ where
         opts: AppendOpts<'a>,
     ) -> Result<Multihash, AppendError> {
         if opts.validation == AppendValidation::Full {
-            self.validate_append_prev_block_exists(&block).await?;
-            self.validate_append_sequence_numbers_integrity(&block)
-                .await?;
-            self.validate_append_seed_block_order(&block)?;
-            self.validate_append_system_time_is_monotonic(&block)
+            let hash = if let Some(b) = opts.precomputed_hash {
+                Cow::Borrowed(b)
+            } else {
+                let hash = self.meta_block_repo.get_block_hash(&block).int_err()?;
+
+                Cow::Owned(hash)
+            };
+            let hashed_block = (hash.as_ref(), &block);
+
+            let validators: StackVisitorsWithDecisionsMutRef<AppendError> = &mut [
+                (
+                    Decision::Stop,
+                    &mut ValidateSeedBlockOrderVisitor::new(hashed_block),
+                ),
+                (
+                    Decision::Stop,
+                    &mut ValidatePrevBlockExistsVisitor::new(hashed_block),
+                ),
+                (
+                    Decision::Stop,
+                    &mut ValidateSequenceNumbersIntegrityVisitor::new(hashed_block),
+                ),
+                (
+                    Decision::Stop,
+                    &mut ValidateSystemTimeIsMonotonicVisitor::new(hashed_block),
+                ),
+                (
+                    Decision::Stop,
+                    &mut ValidateWatermarkIsMonotonicVisitor::new(hashed_block),
+                ),
+                (
+                    Decision::Stop,
```

Comment:
It's good that we got rid of the allocation per visitors, but I don't see why we should ruin ergonomics of the API by requiring users to pass `(Decision, &mut Visitor)` tuples.

One allocation for `Vec<Decision>::with_capacity(visitors.len())` is tolerable. If we wanted to avoid allocations completely (which is IMO overkill) - we could use a stack buffer and overflow to `Vec`.

The point is - let's keep this matter internal to `accept()` and not burden the API users.

Reply:
`MetadataChainImpl::accept()` interface has been simplified -- Done